yarpl/include/yarpl/flowable/Subscriber.h

```diff
@@ -15,40 +15,118 @@ namespace flowable {
 template <typename T>
 class Subscriber : public virtual Refcounted, public yarpl::enable_get_ref {
  public:
+  virtual void onSubscribe(Reference<Subscription>) = 0;
+  virtual void onComplete() = 0;
+  virtual void onError(folly::exception_wrapper) = 0;
+  virtual void onNext(T) = 0;
+};
+
+// codemod all things that inherit from Subscriber to inherit from
+// InternalSubscriber
+template <typename T>
+class InternalSubscriber : public Subscriber<T> {
+ public:
   // Note: If any of the following methods is overridden in a subclass, the new
   // methods SHOULD ensure that these are invoked as well.
-  virtual void onSubscribe(Reference<Subscription> subscription) {
+  void onSubscribe(Reference<Subscription> subscription) override {
     DCHECK(subscription);
-
-    if (subscription_) {
-      subscription->cancel();
-      return;
-    }
-
+    CHECK(!subscription_);
     subscription_ = std::move(subscription);
   }
 
   // No further calls to the subscription after this method is invoked.
-  virtual void onComplete() {
+  void onComplete() override {
     DCHECK(subscription_) << "Calling onComplete() without a subscription";
     subscription_.reset();
   }
 
   // No further calls to the subscription after this method is invoked.
-  virtual void onError(folly::exception_wrapper) {
+  void onError(folly::exception_wrapper) override {
     DCHECK(subscription_) << "Calling onError() without a subscription";
     subscription_.reset();
   }
 
-  virtual void onNext(T) = 0;
-
  protected:
-  Subscription* subscription() {
-    return subscription_.operator->();
+  Reference<Subscription> subscription() {
```

Comment:
I thought we would not have any reference to `subscription_`, but we would just call the cancel & request using a pure virtual method. @alexmalyshev, @lehecka